tsl/src/bgw_policy/job.c

```diff
@@ -379,6 +381,92 @@ bgw_policy_job_check_enterprise_license(BgwJob *job)
 	return required;
 }
 
+static void
+job_execute_function(FuncExpr *funcexpr)
+{
+	bool isnull;
+
+	EState *estate = CreateExecutorState();
+	ExprContext *econtext = CreateExprContext(estate);
+
+	ExprState *es = ExecPrepareExpr((Expr *) funcexpr, estate);
+	ExecEvalExpr(es, econtext, &isnull);
+
+	FreeExprContext(econtext, true);
+	FreeExecutorState(estate);
+}
+
+static void
+job_execute_procedure(FuncExpr *funcexpr)
+{
+	CallStmt *call = makeNode(CallStmt);
+	call->funcexpr = funcexpr;
+	DestReceiver *dest = CreateDestReceiver(DestNone);
+	/* we don't need to create proper param list cause we pass in all arguments as Const */
+#ifdef PG11
+	ParamListInfo params = palloc0(offsetof(ParamListInfoData, params));
+#else
+	ParamListInfo params = makeParamList(0);
+#endif
+	ExecuteCallStmt(call, params, false, dest);
+}
+
+static bool
+job_execute(BgwJob *job)
+{
+	Const *arg1, *arg2;
+	bool started;
+	char prokind;
+	Oid proc;
+	Oid proc_args[] = { INT4OID, JSONBOID };
+	List *name;
+
+	if (!IsTransactionOrTransactionBlock())
+	{
+		started = true;
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+	}
+
+	name = list_make2(makeString(NameStr(job->fd.proc_schema)),
+					  makeString(NameStr(job->fd.proc_name)));
+	proc = LookupFuncName(name, 2, proc_args, false);
+
+	prokind = get_func_prokind(proc);
+
+	arg1 = makeConst(INT4OID, -1, InvalidOid, 4, Int32GetDatum(job->fd.id), false, true);
+	arg2 = makeConst(JSONBOID, -1, InvalidOid, -1, JsonbPGetDatum(job->fd.config), false, false);
+
+	FuncExpr *funcexpr = makeFuncExpr(proc,
+									  VOIDOID,
+									  list_make2(arg1, arg2),
+									  InvalidOid,
+									  InvalidOid,
+									  COERCE_EXPLICIT_CALL);
+
+	switch (prokind)
+	{
+		case PROKIND_FUNCTION:
+			job_execute_function(funcexpr);
+			break;
+		case PROKIND_PROCEDURE:
+			job_execute_procedure(funcexpr);
+			break;
+		default:
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED), errmsg("unsupported function type")));
+			break;
+	}
+
+	if (started)
+	{
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+	}
+
+	return true;
```

Comment:
Should it return `started`?

Reply:
i was just adhering to the other execute_functions, it doesnt make much sense and should be removed once all the policies have been migrated, all the errors are handled by exceptions. It means the job run successfully.